src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,621 @@
+#include "ui_BreakpointViewer.h"
+#include "Convert.h"
+#include "CommClient.h"
+#include "OpenMSXConnection.h"
+
+enum TableColumns {
+	ENABLED = 0,
+	WP_TYPE = 1,
+	LOCATION = 2,
+	BP_ADDRESS = 2,
+	WP_REGION = 2,
+	CONDITION = 3,
+	SLOT = 4,
+	SEGMENT = 5,
+};
+
+struct ItemPos {
+	BreakpointViewer::Type type;
+	int row;
+	bool operator==(const ItemPos &ip) const
+	{
+		return (ip.type == type) && (ip.row == row);
+	}
+};
+
+QList<ItemPos> items;
+
+BreakpointViewer::BreakpointViewer(QWidget* parent)
+	: QTabWidget(parent),
+	  ui(new Ui::BreakpointViewer)
+{
+	setupUi(this);
+	bpTableWidget->sortByColumn(BP_ADDRESS, Qt::AscendingOrder);
+	wpTableWidget->sortByColumn(WP_REGION, Qt::AscendingOrder);
+	bpTableWidget->setColumnHidden(WP_TYPE, true);
+	bpTableWidget->resizeColumnsToContents();
+	wpTableWidget->resizeColumnsToContents();
+
+	selector[0] = bpTableWidget;
+	selector[1] = wpTableWidget;
+	userMode = true;
+
+	connect(bpTableWidget, &QTableWidget::itemChanged, this,
+			&BreakpointViewer::changeBpTableItem);
+	connect(wpTableWidget, &QTableWidget::itemChanged, this,
+			&BreakpointViewer::changeWpTableItem);
+}
+
+// TODO: move the createSetCommand to a session manager
+void BreakpointViewer::createBreakpoint(Type type, int row)
+{
+	// qDebug() << "createBreakpoint() called";
+	auto table = selector[type];
+	Breakpoints::Type type2;
+
+	if (type == WATCHPOINT) {
+		auto model = table->model();
+		auto combo = (QComboBox*) table->indexWidget(model->index(row, WP_TYPE));
+		type2 = static_cast<Breakpoints::Type>(combo->currentIndex() + 1);
+	} else {
+		type2 = Breakpoints::BREAKPOINT;
+	}
+
+	// qDebug() << "createBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
+	// 	qDebug() << "item" << item.type << item.row;
+	// }
+
+	bool ok;
+	int address, endRange;
+	QString location = table->item(row, LOCATION)->text();
+	ok = processLocationField(-1, type, location, address, endRange);
+	if (!ok) {
+		auto enabledItem = table->item(row, ENABLED);
+		userMode = false;
+		enabledItem->setCheckState(Qt::Unchecked);
+		userMode = true;
+		return;
+	}
+
+	QString condition = table->item(row, CONDITION)->text();
+
+	qint8 ps, ss;
+	ok = processSlotField(-1, table->item(row, SLOT)->text(), ps, ss);
+	assert(ok);
+
+	qint16 segment;
+	ok = processSegmentField(-1, table->item(row, SEGMENT)->text(), segment);
+	assert(ok);
+
+	const QString cmdStr = Breakpoints::createSetCommand(type2, address, ps, ss, segment,
+			endRange, condition);
+	auto command = new Command(cmdStr,
+		[this] (const QString result) {
+			emit contentsUpdated(false);
+		}
+	);
+	CommClient::instance().sendCommand(command);
+}
+
+// TODO: move the createRemoveCommand to a session manager
+void BreakpointViewer::replaceBreakpoint(Type type, int row)
+{
+	// qDebug() << "replaceBreakpoint() called";
+	int index = items.indexOf({type, row});
+	// qDebug() << "replaceBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
```

Comment:
(detail)
You probably want:
```   for (const auto& item : items) {```
So `const auto& item` instead of `auto item`, no need to make a copy.

(very much a detail)
I prefer a space at both sides around the ':'. Because here ':' is an operator (like '+') and we put spaces at both sides around operators. In other contexts (e.g. after a label) ':' is a delimiter (like ',') and we only put spaces after a delimiter.